### PR TITLE
Kashirin Alexander Lab 3 Var 1 

### PR DESCRIPTION
**Pull request by AlexanderVladislavovich**

### CODE REVIEW

File: llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+#    __m128d foo2 = _mm_setzero_pd();
+#    return tmp + c;
+#}
+#
+#__m128d muladd5(__m128d a, __m128d b, __m128d c) {
+#    __m128d tmp = a * b * c;
+#    return tmp + c;
+#}
+#
+
```
Comment (m-ly4) [inline llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir:40]: Could you please also support the following case:
```suggestion
#__m128d muladd6(__m128d a, __m128d b, __m128d c) {
#    __m128d tmp = a * b;
#    return c + a * b;
#}

```

Comment (m-ly4) [inline llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir:40]: And add a test case for the following example:
```
#__m128d muladd7(__m128d a, __m128d b, __m128d c) {
#    __m128d tmp = a * b;
#    __m128d tmp2 = tmp + c;
#    return tmp + tmp2;
#}
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir (hunk 1 above)
Comment (AlexanderVladislavovich) [reply llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir:40]: case added

Comment (AlexanderVladislavovich) [reply llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir:40]: case added

File: llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+  bb.0 (%ir-block.3):
+    liveins: $xmm0, $xmm1, $xmm2
+  
+    %2:vr128 = COPY $xmm2
+    %1:vr128 = COPY $xmm1
+    %0:vr128 = COPY $xmm0
+    %3:vr128 = nofpexcept MULPDrr %0, %1, implicit $mxcsr
+    %4:vr128 = nofpexcept ADDPDrr %3, %2, implicit $mxcsr
+    %5:vr128 = nofpexcept ADDPDrr %3, %4, implicit $mxcsr
+    ; CHECK : %4:vr128 = VFMADD213PDr %0, %1, %2, implicit $mxcsr
```
Comment (m-ly4) [inline llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir]: After you replaced
```
    %3:vr128 = nofpexcept MULPDrr %0, %1, implicit $mxcsr
    %4:vr128 = nofpexcept ADDPDrr %3, %2, implicit $mxcsr
```
with
```
%4:vr128 = VFMADD213PDr %0, %1, %2, implicit $mxcsr
```
your pass removed instruction for `%3`. But it is still used by `%5`.

File: llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+body:             |
+  bb.0 (%ir-block.3):
+    liveins: $xmm0, $xmm1, $xmm2
+  
+    %2:vr128 = COPY $xmm2
+    %1:vr128 = COPY $xmm1
+    %0:vr128 = COPY $xmm0
+    %3:vr128 = nofpexcept MULPDrr %0, %1, implicit $mxcsr
+    %4:vr128 = nofpexcept ADDPDrr %3, %2, implicit $mxcsr
+    ; CHECK : %2:vr128 = VFMADD213PDr %0, %0, %0, implicit $mxcsr
```
Comment (m-ly4) [inline llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir]: This is equivalent to code `a*a + a` in source code which is not equal to the initial sample.

File: llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+machineFunctionInfo: {}
+body:             |
+  bb.0 (%ir-block.3):
+    liveins: $xmm0, $xmm1, $xmm2
+  
+    %2:vr128 = COPY $xmm2
+    %1:vr128 = COPY $xmm1
+    %0:vr128 = COPY $xmm0
+    %3:vr128 = nofpexcept MULPDrr %0, %1, implicit $mxcsr
+    %4:vr128 = nofpexcept ADDPDrr %3, %2, implicit $mxcsr
```
Comment (m-ly4) [inline llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir]: Where is check?

File: llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+  bb.0 (%ir-block.3):
+    liveins: $xmm0, $xmm1, $xmm2
+  
+    %2:vr128 = COPY $xmm2
+    %1:vr128 = COPY $xmm1
+    %0:vr128 = COPY $xmm0
+    %3:vr128 = nofpexcept MULPDrr %0, %1, implicit $mxcsr
+    %4:vr128 = nofpexcept MULPDrr %3, %2, implicit $mxcsr
+    %5:vr128 = nofpexcept ADDPDrr %4, %2, implicit $mxcsr
+    ; CHECK : %4:vr128 = VFMADD213PDr %0, %1, %2, implicit $mxcsr
```
Comment (m-ly4) [inline llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir]: Please also add checks for other left instructions in all tests.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir (hunk 2 above)
Comment (AlexanderVladislavovich) [reply llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir]: fixed

File: llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir (hunk 4 above)
Comment (AlexanderVladislavovich) [reply llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir]: check added

File: llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir (hunk 3 above)
Comment (AlexanderVladislavovich) [reply llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir]: fixed

File: llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir (hunk 5 above)
Comment (AlexanderVladislavovich) [reply llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir]: checks added

File: llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir (hunk 2 above)
Comment (m-ly4) [reply llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir]: Could you please revert back input test case and fix pass instead?

File: llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir (hunk 5 above)
Comment (m-ly4) [reply llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir]: I meant to add checks for all instructions of a function like:
```
CHECK:      %3:vr128 = nofpexcept MULPDrr %0, %1, implicit $mxcsr
CHECK-NEXT: %4:vr128 = nofpexcept MULPDrr %3, %2, implicit $mxcsr
CHECK-NEXT: %5:vr128 = nofpexcept ADDPDrr %4, %2, implicit $mxcsr
```

File: llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+body:             |
+  bb.0 (%ir-block.3):
+    liveins: $xmm0, $xmm1, $xmm2
+  
+    %2:vr128 = COPY $xmm2
+    %1:vr128 = COPY $xmm1
+    %0:vr128 = COPY $xmm0
+    %3:vr128 = nofpexcept MULPDrr %0, %1, implicit $mxcsr
+    %4:vr128 = nofpexcept ADDPDrr %3, %2, implicit $mxcsr
+    ; CHECK : %4:vr128 = VFMADD213PDr %0, %1, %2, implicit $mxcsr
```
Comment (m-ly4) [inline llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir]: Need more checks.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir (hunk 2 above)
Comment (AlexanderVladislavovich) [reply llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir]: what exactly should i fix in my pass? should i need to add an additional check for instruction order and make another condition to replace them with fma?

Comment (m-ly4) [reply llvm/test/lab3/kashirin_alexander/X86KashirinMulPass_test.mir]: Your pass should check that the result of Mul operation is not used in other instructions in addition to Add op that you selected for fusion.

File: llvm/lib/Target/X86/lab3/kashirin_alexander/X86KashirinMulPass.cpp [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    MachineBasicBlock &MBB = *MulInstr->getParent();
+    MIMetadata MetaData(*MulInstr);
+
+    BuildMI(MBB, MulInstr, MetaData, TII->get(X86::VFMADD213PDr),
+            AddInstr->getOperand(0).getReg())
+        .addReg(MulInstr->getOperand(1).getReg())
+        .addReg(MulInstr->getOperand(2).getReg())
+        .addReg(AddInstr->getOperand(2).getReg());
+
+    MulInstr->eraseFromParent();
```
Comment (aobolensk) [inline llvm/lib/Target/X86/lab3/kashirin_alexander/X86KashirinMulPass.cpp]: Do we need to check if "Mul" instruction has other uses except this "Add" instruction?

**Review by aobolensk — CHANGES_REQUESTED** [review]

**Review by aobolensk — APPROVED** [review]

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**